Ignore numlock modifier

Danacus opened the pull request:
I ended up dumping the entire modifier map from `xmodmap -pm`. I considered using `xcb::xproto::get_modifier_mapping` similar to `dwm`, but it doesn't really work because I can't access a "key name to key code" mapping from `XcbConnection` (I assume this is by design).

Currently it only ignores `Num_Lock`, but wouldn't be hard to add `Caps_Lock`, because `Caps_Lock` is always `MOD_MASK_LOCK`. The only somewhat tricky thing is having to account for having both `Num_Lock` and `Caps_Lock` enabled as well.

Let me know if there's anything you want me to change before it can be merged.

Danacus committed:
Ignore numlock and capslock

- Very rough first implementation, WIP

Danacus committed:
Get ignored modifier masks from xmodmap

- Added method to 'KeyCode' for ignoring modifiers
- Function to get modifier masks from xmodmap
- Function to generate all possible mask combinations that should be
ignored
- Store the resulting vector of masks in 'XcbConnection'

Danacus committed:
Merge branch 'develop' of https://github.com/sminez/penrose into ignored-modifiers

Danacus committed:
Simplified ignoring modifiers + dump entire modmap from xmodmap

- Added ModMap type
- Changed 'modifiers_from_xmodmap' to dump the entire table instead
- Only ignore Num_Lock for now

sminez commented on src/core/helpers.rs:
```diff
@@ -94,6 +94,49 @@ pub fn keycodes_from_xmodmap() -> CodeMap {
+                    "mod3" => xcb::MOD_MASK_3,
+                    "mod4" => xcb::MOD_MASK_4,
+                    "mod5" => xcb::MOD_MASK_5,
+                    _ => panic!("unknown modifier: {}", name),
```
we probably don't want to panic on an unknown modifier. It should be safe to just drop it, probably with a `warn` level log of the name? That way, if we do drop something that later on someone wants to use, they are able to tell us exactly what it is from the log.

sminez commented on src/core/helpers.rs:
```diff
@@ -94,6 +94,49 @@ pub fn keycodes_from_xmodmap() -> CodeMap {
+
+                    remainder.split(',').for_each(|key| {
+                        // key == <keyname(keycode)>
+                        if let Some(name) = key.split('(').next() {
```
is there an issue with trailing `)` characters being included in this?

Danacus replied:
No, I only care about the part before `(`, the key name. If you wanted to get the keycode from this iterator, you would have to remove the trailing `)`.

sminez commented on src/core/xconnection.rs:
```diff
@@ -520,13 +522,16 @@ impl XcbConnection {
         // xcb docs: https://www.mankier.com/3/xcb_randr_select_input
         xcb::randr::select_input(&conn, root, NOTIFY_MASK).request_check()?;
 
+        let modifier_map = modifiers_from_xmodmap();
```
I'd prefer for the modifiers to be passed through in the parameters to `new` so that we aren't hard coding a dependency on `xmodmap` into `XcbConnection`. The `gen_keybindings` macro is called outside of the constructor for `WindowManager` for the same reason: it allows users to generate the map in an alternate way if they need/want to.

Danacus replied:
That makes sense.

sminez commented on src/core/xconnection.rs:
```diff
@@ -585,9 +590,13 @@ impl XConn for XcbConnection {
-                    })
+                    let mut code = KeyCode::from_key_press(e);
+
+                    if let Some(&numlock_mask) = self.modifier_map.get("Num_Lock") {
```
If we are only using `Num_lock` at this stage, can we just pass that through explicitly rather than grabbing all of the modifiers and keying in to the map every time?

sminez commented on src/core/xconnection.rs:
```diff
@@ -829,17 +838,25 @@ impl XConn for XcbConnection {
     fn grab_keys(&self, key_bindings: &KeyBindings) {
+        let mut modifiers = vec![0];
+
+        if let Some(&numlock_mask) = self.modifier_map.get("Num_Lock") {
```
Is there ever a case where we _wouldn't_ have a NumLock modifier available?

Danacus replied:
I honestly don't know, maybe if someone would intentionally remove `Num_Lock` from `mod2` in some config file?

sminez commented:
@Danacus sorry, what do you mean by 'I can't access a "key name to key code" mapping from XcbConnection'? Are just meaning that the key mappings aren't passed through to `XcbConnection`? If so, I think passing through the modifiers we care about as I've suggested in the review should be fine. We probably don't need/want a dynamic map of modifier names to `xcb` constants as we always care about the same things.

I'm also not so sure now that we do want to ignore caps lock: If you have caps lock enabled, you are definitely sending different key presses through to the OS so we probably want to treat them accordingly. Not sure if the best thing to do is to treat it as if `shift` was held down or to allow for key bindings that rely on `CapsLock` explicitly. (I suppose as a sort of second layer for bindings).

Danacus commented:
I'm not sure about `CapsLock` either. In `dwm` they appear to be ignoring them, but the other options you are suggesting also make sense.

About not needing the dynamic map, perhaps it would be enough to store the modmasks we care about as optional fields in `XcbConnection`. e.g. just add `numlock: Option<u16>` and maybe `capslock: Option<u16>` to `XcbConnection` instead of an entire `HashMap`? We could even set them to `None` in `new` and add another method to set these masks, so we don't break all existing uses of `XcbConnection`.

Danacus committed:
Changed modifier_map to numlock_mask in XcbConnection

- Added set_numlock_mask to XcbConnection
- Updated the simple_config_with_hooks example to demonstrate how this
would work

Danacus commented:
I've quickly demonstrated what I have in mind.

sminez commented on examples/simple_config_with_hooks/main.rs:
```diff
@@ -170,7 +171,13 @@ fn main() {
+    let modifier_map = modifiers_from_xmodmap();
+
+    if let Some(&numlock_mask) = modifier_map.get("Num_Lock") {
+        conn.set_numlock_mask(numlock_mask);
```
I really would prefer this to be a parameter in `new` rather than a method that mutates `XcbConnection` after the fact as mentioned previously.

Danacus committed:
Set numlock_mask in XcbConnection::new

- Removed set_numlock_mask
- Updated all examples

Danacus committed:
Merge branch 'develop' into ignored-modifiers
